src/others/components/map/RegionsSource.tsx

```diff
@@ -0,0 +1,111 @@
+import { useEffect, RefObject } from "react";
+import { useFilter } from "../../contexts/filter";
+import { GeoJSONSource } from "maplibre-gl";
+import { AidRequest } from "../../contexts/api";
+import { adminRegions } from "../../fixtures/regionsP3";
+import { FeatureCollection, GeoJsonProperties, Geometry } from "geojson";
+import { Layer, Source, MapRef } from "react-map-gl";
+import { useDictionaryContext } from "../../contexts/dictionary-context";
+
+
+interface RegionsSourceProperties {
+  aidRequests: AidRequest[] | undefined;
+  mapRef: RefObject<MapRef>;
+  mapLoaded: boolean;
+};
+
+
+export const RegionsSource = ({aidRequests, mapRef, mapLoaded}: RegionsSourceProperties) => {
+  const { translateLocation } = useDictionaryContext();
+  const filterContext = useFilter();
+  const { Dates: dateFilter} = filterContext.filters;
+  const dates = dateFilter?.filterItems.map(({text}) => text) || [];
+
+  const searchParams = new URLSearchParams(window.location.search);
+  const showRegions = (searchParams.get('show_regions') ?
+                       searchParams.get('show_regions') :
+                       process.env.REACT_APP_SHOW_REGIONS) === '1';
+  const regionsSplitByDate = (searchParams.get('regions_split_by_date') ?
+                              searchParams.get('regions_split_by_date') :
+                              process.env.REACT_APP_REGIONS_SPLIT_BY_DATE) === '1';
+
+  const activeDateFilter = filterContext.getActiveFilterItems("Dates")[0];
+  const layerFilterDate = (regionsSplitByDate ?
+                           ["boolean", "true"] :
+                           ["==", ["get", "date"], ["string", activeDateFilter]]);
+
+  const activeCategoryFilters = filterContext.getActiveFilterItems("Categories");
+
+  useEffect(() => {
+    if (!showRegions || !mapRef || !mapRef.current) return;
+    const allRegionsWithMeta = [];
+    for (const date of dates) {
```

Comment:
**issue**

Please don't use loops, they are hard to read and understand, please abide by declarative coding approach. 

Plus, we already have mappers / filters in place - cant't we use the same patterns for mapping the map data?

Additionally, we can't really make any use of Typescript in this code? What is the input type? What is the expected output? 

It would be also great to move all this code out from `useEffect` body to a separate file.

Reply:
At the end of the day - whatever this code does, declaratively or imperatively, it should be tested, typed, and placed in `others/helpers/aid-request-helpers.ts` if it handles aidRequests.

Reply:
replaced all the loops and moved some of the code to the aid-request-helper.ts + added a test